### ISSUE TITLE
Rewrite the rewriter deployment

### ISSUE THREAD

**Issue by renekrie** [querqy/querqy#76]
Rewriter deployment is currently handled as a mix of configuration in solrconfig.xml and additional artefacts (like the 'rules.txt' files for the Common Rules rewriter). 

This approach has a number of issues:

- Applying changes in the configuration requires a collection/core reload. Depending on how long it takes to load the rewriters and on the 'useColdSearcher' setting, Solr can be left without a core to answer queries for a few moments.
- The deployment is just not handy: updating an additional artefact like 'rules.txt' requires pushing data to ZooKeeper in the first step and then reloading the collection in a separate HTTP request. Some Querqy users also had to work around the file size limit in ZooKeeper.

I'd like to suggest rewriting the rewriter configuration and deployment from scratch. The goals should as follows:

- A rewriter should be defined using a single HTTP call (REST API). This call should define, the name under which the rewriter can be used, the rewriter class to be loaded and all the specific settings for this rewriter. This implies that, in the case of the CommonRules rewriter, there will be no separate rules.txt file. (Compare 'Defining a rewriter' in Querqy for Elasticsearch - [https://github.com/querqy/querqy-elasticsearch#defining-a-rewriter ](https://github.com/querqy/querqy-elasticsearch#defining-a-rewriter ))
- The API request handler should take care of validating the rewriter definition, storing it and replacing a previously loaded rewriter instance of the same name in a background process so that search traffic will not be disrupted. The request handler should also make the configuration size handling transparent to the client.
- The API should define means to unload and delete rewriters.
- Rewriter deployment should work in both, SolrCloud and stand-alone mode.

A few thoughts on the implementation:

I think we should get rid of a pre-configured rewriter chain when implementing this instead of creating a second API endpoint to configure the chain. We currently have three elements that are defined at chain level:

1. The list of rewriters. This could be specified per request (see Querqy for Elasticsearch implementation). As a positive side-effect, this would make the chain more flexible.
2. The term query cache. Much as it helped us in older Solr versions (around 4.x), it doesn't seem to provide any performance gain for current Solr/Lucene versions. We might just want to deprecate and remove it.
3. Sinks and mappings for [rewriter logging](https://github.com/querqy/querqy#advanced-configuration-info-logging). I think we can just provide standard sinks that are always available and then configure the mapping at rewriter level.

There seem to be two ideas where to save the rewriter configuration: in ZooKeeper or in a separate collection (maybe Solr's blob store). I leave this as a matter for discussion. If we did it in ZK we would need to make sure that we work around ZK file size limits (splits, compression) and that we are not mis-using ZK. If we used a separate collection we would probably have to think about the order in which collections are loaded. In both cases, we would have to think about how we propagate rewriter update events across nodes.

I've talked with a number of people about this issue, pinging you to join the discussion here: @JohannesDaniel @mkr @MighTguY @tboeghk

**Comment by JohannesDaniel** [comment]
Hi,

I would not worry too much about ZooKeeper. In my experience, ZooKeeper has (sometimes) problems with large files, but not with many files. I have seen Hadoop clusters, where every app (Hadoop, Hive, Spark, Kafka, NiFi, Solr, ...) coordinates its nodes and app-configs via the same ZK ensemble in subdirectories. 

I think I have not seen the case "a collection depends on another collection" at all so far. The additional suggester / spellcheck collections all require a path to a local index to be configured. I think enforcing an order of collections to be loaded might be tricky to be done via plugin? I wonder whether some kind of lazy loading (e. g. triggered by a warmup query) might be easier to be implemented. Another approach might be that (re)loading rules from an existing collection will also include dumping all rules to a local file as a fallback. 

I think the ZK approach should be less hacky, but also limited for future feature enhancements. This approach will not allow to add/delete/update single rules or to search/filter for rules. 

Anyway, this appears to be tough decision :)

**Comment by epugh** [comment]
Have you seen the work on having proper plugins for Solr???   Lots of work
to help abstract you away from just these issues...

I believe that some major components of Solr, like velocity and DIH may
move to this approach and no longer ship with Solr.

It may be too early for this, but when Solr 9 comes out, this will be big
part of it.

**Comment by renekrie** [comment]
@epugh Yes, thank you, I saw it - and forgot to mention it here. As far as I understand the new plugin API just loads the plugin classes in a well defined way -> doesn't help us much with the rewriter configuration.

We should use the plugin API anyway, but it's probably a separate issue.

**Comment by renekrie** [comment]
Thank you, @JohannesDaniel.

I was wondering about the file sizes that would be stored (in many files) in the other systems that you mentioned. I guess a 1000 small files in ZK mighty behave not the same way as 10-20 files of up to 1 MB....
> Hi,
> 
> I would not worry too much about ZooKeeper. In my experience, ZooKeeper has (sometimes) problems with large files, but not with many files. I have seen Hadoop clusters, where every app (Hadoop, Hive, Spark, Kafka, NiFi, Solr, ...) coordinates its nodes and app-configs via the same ZK ensemble in subdirectories.
> 
> I think I have not seen the case "a collection depends on another collection" at all so far. The additional suggester / spellcheck collections all require a path to a local index to be configured. I think enforcing an order of collections to be loaded might be tricky to be done via plugin? I wonder whether some kind of lazy loading (e. g. triggered by a warmup query) might be easier to be implemented. Another approach might be that (re)loading rules from an existing collection will also include dumping all rules to a local file as a fallback.
> 
> I think the ZK approach should be less hacky, but also limited for future feature enhancements. This approach will not allow to add/delete/update single rules or to search/filter for rules.
> 
> Anyway, this appears to be tough decision :)

**Comment by JohannesDaniel** [comment]
> Thank you, @JohannesDaniel.
> 
> I was wondering about the file sizes that would be stored (in many files) in the other systems that you mentioned. I guess a 1000 small files in ZK mighty behave not the same way as 10-20 files of up to 1 MB....
> 

Well, the Hadoop ecosystem files were not too big, of course. And I cannot give you a guarantee that the ZK is going to work as expected. However, over some time we created a new config set via Solr Config API even for smaller rule changes. We temporarily had more than 100 config folders in our ZK (on DEV). We never had any error / warn log entries in the ZK logs or problems in Solr. Again, I would not be worried about the file management in ZK. I think the most critical disadvantage of ZK is the inflexibility regarding additional features to manage rules.

Finally, I was just thinking about creating a ZK ensemble on our K8 and to create, update and read a lot of files in order to see whether sync issues can be observed. Do you think this would help to get a better feeling for ZK file management?

**Comment by mkr** [comment]
I like the idea of configuration change without core reload and the more dynamic rewriter selection. Especially the core reload (or distributed collection reload in SolrCloud) can be painful. It seems to be a “stop-the-world” event, can have unintended side effects (unwanted IW-Commits when the current core is closed etc.) and always disrupts caching.

I like the REST approach of GET/PUT/DELETEing rewriters but wonder whether one then needs to take extra caution that a search application which relies on the existence of certain rewriters (by their symbolic name) can still work gracefully in their absence (e.g. by someone accidentally(?) renaming/deleting a rewriter in a live environment). Currently, in most systems I've seen this step required some kind of software build/deployment with all its pros (tests!) and cons (time + resources). Do we agree that any code changes (i.e. adjustments to rewriter code) will still need software deployment outside the scope of this issue? My general impression of Solrs "Package Manager" (https://lucene.apache.org/solr/guide/8_4/package-manager.html) is that it's more catered for the code deployment rather than the config deployment use case.

I think using Zookeeper and Lucene/Solr's provided ResourceLoader abstraction might be easier to use than a dedicated collection to store configuration. Still a bit worried about ZK's limitations. I remember previous projects where the ZK ensemble became unresponsive when we had znodes with >1000 children. The canonical way was: Announce configuration metadata via ZK but load the configuration (tens of MB) from a shared HA file store (e.g. S3). I would be fine with just GZipping the configuration data (assume we are talking about a few hundred bytes to 100KB max compressed) and store one config blob per rewriter in separate znodes.

We can try the "collection store" route and might have a look at how the BlobStore API does it but I guess all in all it's a bit more effort as you have to control the full life-cycle of that collection (create, schema, configset, adjust replication factor to have it present on *all* nodes etc.).

**Comment by renekrie** [comment]
Thank you, @mkr and @JohannesDaniel for sharing your thoughts.

@mkr - I think it will be okay to leave the responsibility for using correct rewriter names/assuring rewriter existence to the client of Solr/Querqy. In the end, we've heard this wish to be able to configure rewriters via a HTTP/REST API numerous times and there would always be a chance to mess things up in the configuration. 
 
I'm getting more and more convinced that ZK is the right approach. ZK is where we are storing configurations today and it is the place in Solr where configurations should be stored. Last but not least, we'd need ZK anyway even if we used a collection to inform nodes about rewriter configuration updates and, finally, we could go with ZK first and still move to the collection approach later as this would be hidden behind the API. 

@JohannesDaniel please feel free to go ahead with the experiments you mentioned. I tried to find some clues about limitations in ZK but wasn't very successful. This might have some interesting insights: [https://stackoverflow.com/questions/29791134/zookeeper-max-number-of-children-per-node](https://stackoverflow.com/questions/29791134/zookeeper-max-number-of-children-per-node) - related to @mkr's observation on nodes with many children. I think would not hit these sizes but we might want to consider node name lengths (as described in the post on Stackoverflow).

**Comment by mms-jope** [comment]
as we somehow lost the drive here: probably it might be a nice first step simply to think about how to change rules / redeploy without core reload and annoying timeouts.

**Comment by JohannesDaniel** [comment]
as we somehow lost the drive here: probably it might be a nice first step simply to think about how to change rules / redeploy without core reload and annoying timeouts.

**Comment by renekrie** [comment]
@JohannesDaniel I think this is already the most tricky part. Rules are specific to the common rules rewriter. As a rewriter cannot receive an update we will have to change the rewriter framework so that a single rewriter can have a more flexible deployment. In addition, updating the rules (= the rewriter config) means that we need to create a new SolrSearcher in order to invalidate the query caches.

(Re timeouts: do you see timeouts if you set `useColdSearcher=true`)

**Comment by JohannesDaniel** [comment]
@renekrie timeouts are about indexing, collection modification & collection reload...

**Comment by renekrie** [comment]
The above commit is a first cut to configure rewriters via HTTP.

The simplest configuration for Querqy will now be (solrconfig.xml):

```xml

<searchComponent name="query" class="querqy.solr.QuerqyQueryComponent"/>

<requestHandler name="/querqy/rewriter" class="querqy.solr.QuerqyRewriterRequestHandler" />

<queryParser name="querqy" class="querqy.solr.QuerqyDismaxQParserPlugin" />

```

This will already allow to execute search requests using the Querqy query parser via `defType=querqy`.

Rewriters can be configured using URL path `http://mysolr/solr/<collection or core>/querqy/rewriter/<rewriter id>`. The ultimate goal is to use REST semantics (PUT, DELETE, ...) when making requests against this endpoint. For now, we mimic REST predicates by making POST requests and appending `/_put` or `/_delete` to the URL because v1/standalone Solr only allows for POST and GET requests.

(TODO: add proper REST endpoints and delegate to the same methods that handle POSTs against _put and _delete. We'll keep _put and _delete additionally to cover standalone Solr and early Solr 8.x versions. Optional: add `GET /querqy/rewriter/<rewriter id>` to read the configuration and `GET /querqy/rewriter` to get the list of rewriters).

With this commit, a rewriter can thus be configured using  a request like this:

`POST  http://mysolr/solr/<collection or core>/querqy/rewriter/commonrules1/_put` with payload:

```JSON
{
    "class": "querqy.solr.rewriter.commonrules.CommonRulesRewriterFactory",
    "config": {
        "rules" : "notebook =>\nSYNONYM: laptop"
    }
}
```

Except for the Solr-specific class name, the configuration is identical to the format we use in Elasticsearch. [https://docs.querqy.org/querqy/rewriters.html#list-of-available-rewriters](https://docs.querqy.org/querqy/rewriters.html#list-of-available-rewriters).

Updates to rewriter configurations are applied without core/collection reloads. Configurations are validated before saving. An invalid configuration will not be saved and, if available, the previous, valid configuration will be continued to be used.   

To apply one or more rewriters, they must be listed in the new request parameter `rewriters`, for example: 

`q=notebook&defType=querqy&rewriters=replace,commonrules1,wordbreak`

Rewriters are applied in the order of their occurrence in the parameter value. This means that the old rewrite chain - that was set at the query parser in solrconfig.xml - is now replaced by this parameter, which is much more flexible as it can be set per request. It will probably reduce the configuration size in solrconfig.xml, especially in multi-tenancy setups.

As this change makes breaking changes to the configuration anyway, I've renamed a number of classes and packages that are exposed to the Querqy user. I've also removed classes that we don't really support (like the Lucene AnalysisQueryParser). The most important refactorings include:

* rename `querqy.solr.contrib` to `querqy.solr.rewriter` - all 'contribs' are rewriters and we fully support them
* create a sub-package per rewriter under `querqy.solr.rewriter` if a rewriter has more than one class
* get rid of names starting with 'Default...' or 'Simple...' - we are using just one implementation of the corresponding component anyway.

Given that this issue involves many breaking changes in the configuration, I suggest that we increase the major version number of querqy-for-lucene (to v5).



*Notes on implementation*

The core component is `querqy.solr.QuerqyRewriterRequestHandler`. It exposes rewriter management via HTTP and via the internal RequestHandler name. 

The QuerqyDismaxQParserPlugin and the TermQueryCachePreloader Solr event listener both access rewriters via the RequestHandler. The RequestHandler is loaded before the QParserPlugin and the Solr event listener. When it is loading, it checks the configset in ZK or the conf folder in standalone Solr for rewriter configurations and loads the rewriter factories as they are defined in these configurations.

The loading and caching of rewriter factories is encapsulated in the `RewriterContainer`. The rewriter container also implements methods for saving and deleting rewriters. As this needs to be implemented differently for SolrCloud and standalone Solr, the RewriterContainer uses a template method approach with specific implementations for the two cases (`ZkRewriterContainer` vs `StandaloneRewriterContainer`) ZkRewriterContainer also implements ZK watchers so that rewriter configuration updates will be propagated to all nodes in the Solr cluster.

In the SolrCloud version, rewriter configurations are gzipped for storage in ZK. (TODO: automatically split files > 1 MB to avoid hitting the ZK default size limit).

TODO: I've tried to keep the QuerqyRewriterRequestHandler a zero-config component. Maybe this is not necessary - I don't know enough yet about how well configuring the request handler will work using the new plugin API and the v2 API plugin capabilities. If it's easy to configure, we could think about managing (or just updating) the term query cache ([https://docs.querqy.org/querqy/solr-plugin-configuration.html#term-query-cache](https://docs.querqy.org/querqy/solr-plugin-configuration.html#term-query-cache)) ourselves from within the QuerqyRewriterRequestHandler (or the RewriterContainer) instead of relying on a separately configured SolrEventListener (the `TermQueryCachePreloader`). We would simply register the QuerqyRewriterRequestHandler as a Solr event listener and reload the term query cache when the event fires - however, the RequestHandler configuration would need to tell us so.

As already mentioned, the configuration looks very similar between Solr and ES, including the configuration validation method of the rewriter factories. We could think of providing a single implementation for the two cases but I'm hesitant to tie the two together too much.

I've started creating rewriter configuration request builders (`RewriterConfigRequestBuilder`, `querqy.solr.rewriter.commonrules.CommonRulesConfigRequestBuilder`, `querqy.solr.rewriter.replace.ReplaceConfigRequestBuilder`). They allow to set configuration properties using a 'meaningful' object model and also allow to read in resources from an input stream (for example, to recycle the old rules.txt files). I've started this more or less to make it easier for me to write tests but we might want to extend this to a nice configuration client support, including methods for building SYNONYM rules etc. for the CommonRulesRewriter.

**Comment by renekrie** [comment]
As of a discussion with @JohannesDaniel: to facilitate the cold start, there should be a toggle that controls the handling of rewriters that are specified in request url parameter `rewriters=` but that are missing in the configuration. If that toggle is on, missing rewriters will be skipped and an error message will be logged. If the toggle is off, a 'bad request' error will be returned. The default should be 'off'.

**Comment by tkaessmann** [comment]
Just added the request handler to get all rewriters as a pull request (https://github.com/querqy/querqy/pull/132).
A very helpful curl to add a rewriter at the current stage was:
```CURL
curl --request POST 'http://localhost:8983/solr/CORE/querqy/rewriter/commonrules1?action=save' \
--data-raw '{
    "class": "querqy.solr.rewriter.commonrules.CommonRulesRewriterFactory",
    "config": {
        "rules" : "notebook =>\nSYNONYM: laptop"
    }
}'
```

**Comment by renekrie** [comment]
The Solr testing framework can decide to have no live nodes in the cloud client -> I changed our tests to use HttpClients. I also made sure the rewriter cleanup after tests waits until the deletes have been propagated to the nodes. a67202c

**Comment by renekrie** [comment]
Implemented and merged (PR #155)
